src/main/kotlin/com/exactpro/th2/http/client/Main.kt

```diff
@@ -115,105 +149,199 @@ fun main(args: Array<String>) = try {
+    val onRequest: (RawHttpRequest) -> Unit
+    val onResponse: (RawHttpRequest, RawHttpResponse<*>) -> Unit
 
-        messageRouter.send(rawMessage.toBatch(), SECOND.toString())
-        eventRouter.storeEvent(
-            Event
-                .start()
-                .name("Sent HTTP request"),
-            if (rawMessage.hasParentEventId()) rawMessage.parentEventId else rootEventId
-        )
-    }
+    val executor = Executors.newSingleThreadScheduledExecutor()
+    registerResource("message batch executor") { executor.shutdownGracefully() }
 
-    val onResponse = { request: RawHttpRequest, response: RawHttpResponse<*> ->
-        messageRouter.send(response.toRawMessage(connectionId, incomingSequence(), request).toBatch(), FIRST.toString())
-        stateManager.onResponse(response)
-    }
+    with(settings) {
+        val book = rootEventId.bookName
+        val sessionGroup = sessionAlias
 
-    val client = HttpClient(
-        settings.https,
-        settings.host,
-        settings.port,
-        settings.readTimeout,
-        settings.keepAliveTimeout,
-        settings.maxParallelRequests,
-        settings.defaultHeaders,
-        stateManager::prepareRequest,
-        onRequest,
-        onResponse,
-        stateManager::onStart,
-        stateManager::onStop,
-        settings.validateCertificates,
-        settings.certificate
-    ).apply { registerResource("client", ::close) }
-
-    stateManager.runCatching {
-        registerResource("state-manager", ::close)
-        init(StateManagerContext(client, settings.auth))
-    }.onFailure {
-        LOGGER.error(it) { "Failed to init state manager" }
-        eventRouter.storeEvent(rootEventId, "Failed to init state manager", "Error", it)
-        throw it
-    }
+        val eventBatcher = EventBatcher(
+            maxBatchSizeInItems = maxBatchSize,
+            executor = executor,
+            maxFlushTime = maxFlushTime,
+            onBatch = eventRouter::send
+        ).also { registerResource("event batcher", it::close) }
 
-    requestHandler.runCatching {
-        registerResource("request-handler", ::close)
-        init(RequestHandlerContext(client))
-    }.onFailure {
-        LOGGER.error(it) { "Failed to init request handler" }
-        eventRouter.storeEvent(rootEventId, "Failed to init request handler", "Error", it)
-        throw it
-    }
+        val onError: (Throwable) -> Unit = {
+            eventBatcher.storeEvent(rootEventId, "Batching problem: ${it.message}", "Message batching problem", it)
+        }
+
+        if (useTransport) {
+            val messageBatcher = if (batchByGroup) {
+                MessageBatcher(maxBatchSize, maxFlushTime, book, GROUP_SELECTOR, executor, onError, transportMR::send)
+            } else {
+                MessageBatcher(maxBatchSize, maxFlushTime, book, ALIAS_SELECTOR, executor, onError) {
+                    transportMR.send(it, it.groups.first().messages.first().id.direction.proto.toString())
```

Comment:
Why do we use direction here? It won't work with the new cradle, will it? There should not be overlapping between batches in one group but if we group messages by alias there will be overlapping.